drivers/can/can_esp32_twai.c

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) 2022 Henrik Brix Andersen <henrik@brixandersen.dk>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT espressif_esp32_twai
+
+#include "can_sja1000.h"
+
+#include <zephyr/drivers/can.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/interrupt_controller/intc_esp32.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/logging/log.h>
+
+#include <soc.h>
+
+LOG_MODULE_REGISTER(can_esp32_twai, CONFIG_CAN_LOG_LEVEL);
+
+struct can_esp32_twai_config {
+	mm_reg_t base;
+	const struct pinctrl_dev_config *pcfg;
+	const struct device *clock_dev;
+	const clock_control_subsys_t clock_subsys;
+	int irq_source;
+};
+
+static uint8_t can_esp32_twai_read_reg(const struct device *dev, uint8_t reg)
+{
+	const struct can_sja1000_config *sja1000_config = dev->config;
+	const struct can_esp32_twai_config *twai_config = sja1000_config->custom;
+	mm_reg_t addr = twai_config->base + reg * sizeof(uint32_t);
+
+	return sys_read32(addr) & 0xFF;
+}
+
+static void can_esp32_twai_write_reg(const struct device *dev, uint8_t reg, uint8_t val)
+{
+	const struct can_sja1000_config *sja1000_config = dev->config;
+	const struct can_esp32_twai_config *twai_config = sja1000_config->custom;
+	mm_reg_t addr = twai_config->base + reg * sizeof(uint32_t);
+
+	sys_write32(val & 0xFF, addr);
+}
+
+static int can_esp32_twai_get_core_clock(const struct device *dev, uint32_t *rate)
+{
+	ARG_UNUSED(dev);
+
+	/* The internal clock operates at half of the oscillator frequency */
+	*rate = APB_CLK_FREQ / 2;
```

Comment:
Question, this APB_CLK_FREQ should not be get using some function from clock control device driver?

Reply:
Ideally, yes. There is no ESP32 clock controller driver allowing to retrieve this clock rate. Other in-tree ESP32 drivers requiring this clock rate also use the define.